### PR TITLE
Login test

### PR DESCRIPTION
**Pull request by luciagonzalezdm**
Nano-experiencia 1

### CODE REVIEW

File: tests/login.test.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
+describe('DigitalBank', () => {
+
+    //example test
+    it('Should enter the site', async () => {
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Consider using, whenever possible, a naming structure that emphasizes what **should** happen **when** some action is performed / the application is in a given state. This makes it easy to figure out what to expect from the test without seeing its implementation.

Eg. Something like: "**Should** display login page **when** browsing to the `/login` route"

File: tests/login.test.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      browser.url('login');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: This test needs an assertion, as in its current state, it's not really validating anything from the page, except commanding the webdriver to browse a given page.

Add an assertion that checks for a condition that helps the test figure out if expected conditions are being met.
Eg.

```javascript
browser.url('login');
expect(browser.title().value()).toMatch('login-page');
```

This last **expect** line will fail the test if that title is not found on the page it's browsing (That is an example of an element you can look at to determine success of an assertion, you can check for any other element that signals that you are browsing the right page)

File: tests/login.test.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    });
+
+    it('Should log in to the demo account', async () => {
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same comment regarding using a **should / when** based structure for test names

File: tests/login.test.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      browser.url('login');
+      
+      let username = await $('[id="username"]');
+      await username.setValue('jsmith@demo.io');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Use **page objects** to abstract out logic regarding elements location / interaction, then make the tests interact with pages exclusively through them. That way you decouple low-level details, such as locating elements, calling the right webdrvierio method on them, etc, from actual test definition.

Ideally the flow will resamble something like this, after you create the loginPage page object:

```javascript
loginPage.open('login');
loginPage.logIn('jsmith@demo.io', 'Demo123!');
```

Basically encapsulating the complexities of loging in into methods with higher-level names

See: https://webdriver.io/docs/pageobjects/

File: tests/login.test.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+       let password = await $('[id="password"]');
+      await password.setValue('Demo123!');
+
+      let loginButton = $('[id="submit"]');
+      await loginButton.click()
+      await browser.pause(5000);
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: This is a line only needed for debugging purposes. Delete it, as it adds an unnecessary delay between tests.

File: tests/login.test.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      browser.url('login');
+      
+      let username = await $('[id="username"]');
+      await username.setValue('jsmith@demo.io');
+
+       let password = await $('[id="password"]');
+      await password.setValue('Demo123!');
+
+      let loginButton = $('[id="submit"]');
+      await loginButton.click()
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: As noted in a previous test, in order to be complete, tests need an assertion clause (could use chai's **assertion** or **expect** function, but prefer **expect** because of readability reasons) to explicitly express what do you expect from the final state you leave the application after applying all those commands in order to signal that the right condition has been met.

File: tests/login.test.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      await browser.pause(5000);
+
+    });
+
+    it('Should log in to the demo account using enter key', async () => {
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same comment regarding naming convention for tests

File: tests/login.test.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      browser.url('login');
+      
+      let username = await $('[id="username"]');
+      await username.setValue('jsmith@demo.io');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same comment regarding using page objects to encapsulate low-level details of element interaction

File: tests/login.test.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+       let password = await $('[id="password"]');
+      await password.setValue('Demo123!');
+
+      let loginButton = $('[id="submit"]');
+      await loginButton.keys('Enter');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same comment regarding using an **expect** clause at the end of the test to explicitly assert an expected condition

File: tests/login.test.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      await browser.pause(5000);
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Remove debugging code

File: tests/login.test.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+    });
+
+    it('Should deny access and show error message', async () => {
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Make sure to add a **when** sentence to describe what needs to happen in order to obtain that behaviour (that you correctly defined in your **should** sentence)

File: tests/login.test.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      browser.url('login');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same general comments than in previous tests

File: tests/login.test.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      let loginButton = $('[id="submit"]');
+      await loginButton.click();
+      await browser.pause(5000);
+
+    });
+
+    it('Remember Me button should work', async () => {
+      browser.url('login');
+      
+      let username = await $('[id="username"]');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same comments than in previous tests

File: tests/DOCUMENTATION.md [hunk 14]
```diff
@@ -0,0 +1,36 @@
+** #Documentation **
```
Comment (rquintana-abstracta) [inline tests/DOCUMENTATION.md]: The name of this file should better reflect the topic of this documentation. Something that points to test cases definition, or the like. `DOCUMENTATION.md` is too broad

File: package.json [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    "url": "git+https://github.com/luciagonzalezdm/nanoexp-digitalbank.git"
+  },
+  "author": "luciagonzalez",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/luciagonzalezdm/nanoexp-digitalbank/issues"
+  },
+  "homepage": "https://github.com/luciagonzalezdm/nanoexp-digitalbank#readme",
+  "devDependencies": {
+    "webdriverio": "^7.16.12"
```
Comment (rquintana-abstracta) [inline package.json:30]: This project needs more libraries. Consider installing mocha, chai and chai-webdriverio (and every library that wdio initially installs when you are configuring the project) again, so you `package.json` reflects them (as this is what a user of your repo will need to reproduce your environment)

File: README.md [hunk 16]
```diff
@@ -1 +1,33 @@
...
+## Prerequisites
+1. Install Java SDK
+2. Install node.js
+3. Node Package Manager (npm)
+
+4. Install Visual Studio Code IDE
+⋅⋅* wdio.config
+⋅⋅* npm install @wdio/cli
+⋅⋅* npm install chai
+⋅⋅* npm install chai-webdriverio
```
Comment (rquintana-abstracta) [inline README.md]: This will not be needed if you have them on your `package.json`, as users only need to do a general `npm install`, when their working directory is your project directory, and they'll be installing the same dependencies you are using.

File: pages/login.page.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    get signInButton () { return $('#submit'); }
+    get rememberMe () { return $('#remember-me'); }
+    get alert () {return $('[class= "badge badge-pill badge-danger"]'); }
+
+
+    async logIn(username, password) {
+        await this.usernameInput.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await this.usernameInput.setValue(username);
+        await this.passwordInput.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await this.passwordInput.setValue(password);
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: There is duplication of the pattern of waiting for an element to be clickable and setting their value here. Consider encapsulating that logic into its own method: Something like `completeField(element, text)` that waits for an element to be clickable and then sends it the string you pass it by parameter. That way you reduce overall lines for `logIn()` method and make it more readable (as you make it work with constructs of higher-level of abstraction).

Also, this `logIn()` method is not actually logging in, it's just filling the form fields (not pressing enter nor clicking the submit button). Make sure you provide implementations for that as well (something like `logInWithEnter()` and `logInWithSubmitButton()`)

File: pages/login.page.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    async open(link) {
+        await browser.url(`${link}`);
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: As you are not interpolating a string here, but rather just passing a raw string directly to `browser.url()`, there's no need for **template literal** syntax. You can just do: `await browser.url(link)` to simplify it

File: tests/login.test.js [hunk 19]
```diff
@@ -0,0 +1,55 @@
+import LoginPage from '../pages/login.page';
+
+describe('DigitalBank', () => {
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: This suite is actually about testing login functionality. This name is too broad. Consider using something that refers to the fact that you are testing login stuff

File: tests/login.test.js [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+      await LoginPage.logIn('jsmith@demo.io', 'Demo123!');
+      await LoginPage.clickElement(LoginPage.signInButton);
+      expect(browser).toHaveUrlContaining('home');
+
+    });
+
+    it('Should log into demo account when using enter key', async () => {
+      await LoginPage.open('login');
+      await LoginPage.logIn('jsmith@demo.io', 'Demo123!');
+      let loginButton = await $('[id="submit"]');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: You are mixing 2 levels of abstraction here (pageObjects methods and raw web elements located through selectors). Ideally, any function should work at the same level of abstraction to make it easier to mantain and understand. Consider implementing this loginButton as a pageObject attribute (the same way you did with username input and password), and access it through one of the logIn methods proposed in a previous comment. That way you encapsulate all the complexity regarding how to interact with page's elements at pageObject level (making the process transparent for the test)

File: tests/login.test.js [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+      await loginButton.keys('Enter');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same observation as with the loginButton location.

File: tests/login.test.js [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+    it('Should remember the session when the button is checked', async () => {
+      await LoginPage.open('login');
+      await LoginPage.logIn('jsmith@demo.io', 'Demo123!');
+      await LoginPage.clickElement(LoginPage.rememberMe);
+      await LoginPage.clickElement(LoginPage.signInButton);
+
+      await browser.close ()
+      await LoginPage.open('home');
+      let welcomeMsg = await $('[class="active"]') ;
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same comment regarding impelmenting this as an attribute of a pageObject (In this case, you'll probably need a HomePage object)

File: pages/login.page.js [hunk 23]
```diff
@@ -0,0 +1,29 @@
+const PAGE_TIMEOUT = 10000;
+
+class LoginPage {
+    
+    get usernameInput () { return $('#username'); }
+    get passwordInput () { return $('#password'); }
+    get signInButton () { return $('#submit'); }
+    get rememberMe () { return $('#remember-me'); }
+    get alert () {return $('[class= "badge badge-pill badge-danger"]'); }
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: This is not a very descriptive name. Firstly, it can be confused with a javascript alert (that is a pop up message that appears at the top of a page typically), secondly, it's a bit too general, as it doesn't hint about its contents: Prefer something that points to its function on the page (like: `failedLoginMessage`) or something like that

File: tests/login.test.js [hunk 24]
```diff
@@ -0,0 +1,55 @@
+import LoginPage from '../pages/login.page';
+
+describe('DigitalBank', () => {
+
+    //example test
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Delete this comment, as it seems like something used for debugging purposes

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement recommended changes

File: README.md [hunk 25]
```diff
@@ -1 +1,42 @@
...
+
+
+## Prerequisites
+1. Install Java SDK
+2. Install node.js
+3. Node Package Manager (npm)
+
+4. Install Visual Studio Code IDE
+- wdio.config
+- npm install @wdio/cli
```
Comment (rquintana-abstracta) [inline README.md]: This is not needed. This readme needs a section to explicitly indicate how to install dependencies for the project (soemthing you achive by simply running `npm install`)

File: pages/home.page.js [hunk 26]
```diff
@@ -0,0 +1,8 @@
+const PAGE_TIMEOUT = 10000;
+
+class HomePage {
+       
+        get welcomeMsg () { return $('[class="active"]'); }
```
Comment (rquintana-abstracta) [inline pages/home.page.js]: This selector seems a bit flaky, as it makes reference to a rather general class that can potentially be used on any other element. Consider using a more robust selector (something that more closely refers to that specific element and that has more potential for unicity )

File: pages/login.page.js [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    get rememberMe () { return $('#remember-me'); }
+    get failedLoginMessage () {return $('[class= "badge badge-pill badge-danger"]'); }
+
+
+    async completeFields(element, text) {
+        await element.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await element.setValue(text);
+    }
+
+    async open(link) {
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: There are methods that you are re-implementing on every page object (thus duplicating them) that are common to any page you are modeling: This is an example (`open()`), `completeFields()` and `clickElement()` are other examples.

Consider implementing behaviour that is common to every class in a super class (In page object lingo, that'd be something like a **BasePage**), and make every other page **inherit** from it, that way you only need to define those methods there, and pages than **inherits / extends** from it automatically get those methods by default, removing that duplication.

File: tests/login.test.js [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+      await LoginPage.open('login');
+      await LoginPage.completeFields(LoginPage.usernameInput, 'jsmith@demo.io');
+      await LoginPage.completeFields(LoginPage.passwordInput, 'Demo123!');
+      await LoginPage.logInWithEnter();
+      expect(browser).toHaveUrlContaining('home');
+
+    });
+
+    it('Should deny access when invalid data is entered', async () => {
+      await LoginPage.open('login');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: I think you can encapsulate this a bit more, as it makes sense that each page knows the URL they are located at, so the test doesn't need to. Consider implementing a method that navigates to the login page (potentially using your current `open()` implementation under the hood). Something like `navigateToPage()`, so when you use it from the test level, the call looks like: `LoginPage.navigateToPage()` (without needing to pass a specific parameter).

`navigateToPage()` method would be implemented in `LoginPage` class like:

```javascript

async navigateToPage(){
    await this.open('login')
}
````

This better applies the principle of **Information Expert** from the **GRASP** design recommendations we discussed last week: https://en.wikipedia.org/wiki/GRASP_(object-oriented_design)

File: tests/login.test.js [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+    });
+
+    it('Should remember the inputs when the button is checked', async () => {
+      await LoginPage.open('login');
+      await LoginPage.completeFields(LoginPage.usernameInput, 'jsmith@demo.io');
+      await LoginPage.completeFields(LoginPage.passwordInput, 'Demo123!');
+      await LoginPage.clickElement(LoginPage.rememberMe);
+      await LoginPage.logInWithSubmitButton();
+
+      await browser.newWindow('login');
```
Comment (rquintana-abstracta) [inline tests/login.test.js:46]: as you use the **login** string twice in this test, it might be safer to declare a constant for it and then use it in those two places

File: tests/login.test.js [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+      console.log(await LoginPage.usernameInput.getValue());
+      
+      expect(LoginPage.usernameInput).toHaveText('jsmith@demo.io');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Same with this, regarding using the `jsmith@demo.io` string twice throughout the test, create a constant for it and pass it to those methods

File: tests/logout.test.js [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+  it('Should close a session when you try to have 2 simultaneously', async () => {
+    await LoginPage.open('login');
+    await LoginPage.completeFields(LoginPage.usernameInput, 'jsmith@demo.io');
+    await LoginPage.completeFields(LoginPage.passwordInput, 'Demo123!');
+    await LoginPage.logInWithSubmitButton();
+
+    await browser.newWindow('login');
+    await LoginPage.completeFields(LoginPage.usernameInput, 'nsmith@demo.io');
+    await LoginPage.completeFields(LoginPage.passwordInput, 'Demo123!');
+    await LoginPage.logInWithSubmitButton();
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: This pattern (filling in fields for login inputs and pressing a button or enter) repeats throughout most test, and, in particular, repeats in this test. This is a good sign that they maybe belong together in a higher-level operation that abstracts them out a bit more. Maybe `loginInWithSubmitButton()` and `logInWithEnter()` should include the completing fields operation, so you pretty much directly use them to fullfill the whole login process: Something like: `LoginPage.logInWithSubmitButton('jsmith@demo.io', 'Demo123!')` essentially parameterizing username and password, and then internally using `completeFields()` and clicking the specific button / pressing the specific key after that. This will remove that duplication, shorten and improve readability of the tests

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement proposed changes

File: tests/login.test.js [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+
+    it('Should remember the inputs when the button is checked', async () => {
+      await LoginPage.open('login');
+      await LoginPage.completeFields(LoginPage.usernameInput, 'jsmith@demo.io');
+      await LoginPage.completeFields(LoginPage.passwordInput, 'Demo123!');
+      await LoginPage.clickElement(LoginPage.rememberMe);
+      await LoginPage.logInWithSubmitButton();
+
+      await browser.newWindow('login');
+      console.log(await LoginPage.usernameInput.getValue());
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Remove this line, as it's for debugging purposes

File: pages/login.page.js [hunk 33]
```diff
@@ -0,0 +1,38 @@
+import BasePage from "./base.page";
+const PAGE_TIMEOUT = 10000;
+const BASE_PAGE = BasePage;
+
+class LoginPage {
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: In order to apply inheritance, you need to make `LoginPage` **extend** from `BasePage`. Full class declaration looks like:

```javascript
class LoginPage extends BasePage
```

That way you gain access to superclass (BasePage in this case) attributes and methods by prefixing your attributes / method calls by a **super**.

File: pages/login.page.js [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+    
+    get usernameInput () { return $('#username'); }
+    get passwordInput () { return $('#password'); }
+    get rememberMe () { return $('#remember-me'); }
+    get logInButton () { return $('#submit'); }
+    get failedLoginMessage () {return $('[class= "badge badge-pill badge-danger"]'); }
+
+    async navigateToPage(){
+        await BASE_PAGE.open('login');
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: After you make this class inherit from BasePage, you gain access to super-class' methods and attributes, and you can call them by prefixing **super** to the method call. In this case, it'll be something like:

```javascript
await super.open('login')
```

**super** holds a reference to the superclass of any class that uses it as part of their method / attributes declaration

See reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super

File: pages/login.page.js [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    
+    async completeCredentials(username, password) {
+        await BASE_PAGE.completeFiels(await this.usernameInput, username);
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: Same comment as above, after inheritance is implemented, you use super instead of BASE_PAGE to access this method

File: pages/login.page.js [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+        await BASE_PAGE.completeFields(await this.passwordInput, password);
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: Same comment regarding **super**

File: pages/login.page.js [hunk 37]
```diff
@@ -0,0 +1,38 @@
...
+        await this.logInButton.keys('Enter');
+    }
+
+    async logInWithSubmitButton(username, password){
+        this.completeCredentials(username, password);
+        await this.logInButton.click();
+    }
+
+    async logInWithRememberMeChecked(username, password){
+        await BASE_PAGE.clickElement(this.rememberMe);
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: Same comment regarding **super**

File: pages/login.page.js [hunk 38]
```diff
@@ -0,0 +1,38 @@
+import BasePage from "./base.page";
+const PAGE_TIMEOUT = 10000;
+const BASE_PAGE = BasePage;
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: This is not needed if you apply inheritance

File: pages/login.page.js [hunk 39]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    
+    async completeCredentials(username, password) {
+        await BASE_PAGE.completeFiels(await this.usernameInput, username);
+        await BASE_PAGE.completeFields(await this.passwordInput, password);
+      }
+
+    async logInWithEnter(username, password){
+        this.completeCredentials(username, password);
+        await this.logInButton.keys('Enter');
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: Prefer implementing a method in BasePage that presses enter on some element, with similar interface than your current `clickElement(element)`, and using it here instead of raw key sending, as a means to remain at the same level of abstraction throughout the whole method. Currently you are mixing 2 levels: credentials handling and web element's direct method calling.

File: pages/login.page.js [hunk 40]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    async logInWithSubmitButton(username, password){
+        this.completeCredentials(username, password);
+        await this.logInButton.click();
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: Prefer using your super class `clickElement()` for this operation, as it already implements an explicit wait and also you use it in other methods (achieving usage consistency).

File: tests/login.test.js [hunk 41]
```diff
@@ -0,0 +1,52 @@
...
+      await LOGIN_PAGE.logInWithSubmitButton();
+
+      expect(LOGIN_PAGE.failedLoginMessage).toHaveText('Error');
+    });
+
+    it('Should remember the inputs when the button is checked', async () => {
+      await LOGIN_PAGE.logInWithRememberMeChecked(VALID_USERNAME, VALID_PASSWORD);
+      await browser.newWindow('login');
+  
+      expect(LOGIN_PAGE.usernameInput).toHaveText('jsmith@demo.io');
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: you have a constant defined that holds this username, use it instead of a raw string

File: tests/logout.test.js [hunk 42]
```diff
@@ -0,0 +1,37 @@
...
+const VALID_PASSWORD = 'Demo123!';
+
+
+describe('Logout Test Cases', () => {
+  beforeEach(async function  () {
+    await LOGIN_PAGE.navigateToPage(); 
+  });
+
+  it('Should logout successful when click on the Logout button', async () => {
+    await LOGIN_PAGE.logInWithSubmitButton(VALID_USERNAME, VALID_PASSWORD);
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: This is a pre-requisite for both tests, so it should be part of the **beforeEach hook** as well

File: tests/logout.test.js [hunk 43]
```diff
@@ -0,0 +1,37 @@
...
+    await BASE_PAGE.clickElement(LogoutPage.profileIcon);
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: After you make LoginPage inherit from BasePage, you can call this method directly on LoginPage. Make sure to change this call to that after you implement inheritance

File: tests/logout.test.js [hunk 44]
```diff
@@ -0,0 +1,37 @@
...
+    await BASE_PAGE.clickElement(LogoutPage.logoutButton);
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: Same comment as above

File: pages/logout.page.js [hunk 45]
```diff
@@ -0,0 +1,13 @@
+const PAGE_TIMEOUT = 10000;
```
Comment (rquintana-abstracta) [inline pages/logout.page.js]: This is not needed anymore in this file

File: pages/base.page.js [hunk 46]
```diff
@@ -0,0 +1,18 @@
...
+        await element.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await element.setValue(text);
+    }
+
+async clickElement (element) {
+        await element.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await element.click();
+    }
+}
+export default new BasePage();
```
Comment (rquintana-abstracta) [inline pages/base.page.js]: As this is an **abstract class**, it's meant to be used **indirectly** by inheriting from it, so you don't **instantiate** an object of type BasePage, you directly export the class such as:

```javascript
export default BasePage

```

This is different than with concrete classes like LoginPage or LogoutPage, that exports an instantiated object

File: pages/logout.page.js [hunk 47]
```diff
@@ -0,0 +1,13 @@
...
+
+class LogoutPage {
```
Comment (rquintana-abstracta) [inline pages/logout.page.js]: Make sure you **extend** from BasePage here as well

File: tests/login.test.js [hunk 48]
```diff
@@ -0,0 +1,52 @@
+import LOGIN_PAGE from '../pages/login.page';
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: Convention naming for page objects in webdriver.io favours **PascalCase** for page objects. Make sure to name this accordingly (LoginPage instead of LOGIN_PAGE, which is a pattern used for constants)

File: tests/login.test.js [hunk 49]
```diff
@@ -0,0 +1,52 @@
...
+const LOGIN_PAGE = LOGIN_PAGE;
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: this is not needed

File: tests/logout.test.js [hunk 50]
```diff
@@ -0,0 +1,37 @@
+import LoginPage from '../pages/login.page';
+import LogoutPage from '../pages/logout.page';
+import HomePage from '../pages/home.page';
+import BasePage from '../pages/base.page';
+
+const LOGIN_PAGE = LoginPage;
+const BASE_PAGE = BasePage;
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: These are not needed, and BasePage won't be directly used at this level, only for other pages to inherit from it

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement new changes

File: pages/login.page.js (hunk 38 above)
Comment (luciagonzalezdm) [reply pages/login.page.js]: Are you referring to the constant declaration or to the import?

Comment (rquintana-abstracta) [reply pages/login.page.js]: The constant, yeah. The import is needed.

File: tests/login.test.js (hunk 48 above)
Comment (luciagonzalezdm) [reply tests/login.test.js]: Thank you! When I used Replace All to change the calls to consts ,I didnt notice that everything had changed

File: pages/base.page.js [hunk 51]
```diff
@@ -0,0 +1,22 @@
...
+async clickElement (element) {
+        await element.waitForClickable({ timeout: PAGE_TIMEOUT });
+        await element.click();
+    }
+
+    async pressEnterKey (element) {
+        await element.keys('Enter');
+    }
+}
+export default BasePage();
```
Comment (rquintana-abstracta) [inline pages/base.page.js]: you have to just export the class, without calling it. Just `export default BasePage`

File: pages/home.page.js [hunk 52]
```diff
@@ -0,0 +1,8 @@
+const PAGE_TIMEOUT = 10000;
+
+class HomePage {
```
Comment (rquintana-abstracta) [inline pages/home.page.js]: Inherit from BasePage here as well

File: pages/home.page.js [hunk 53]
```diff
@@ -0,0 +1,8 @@
+const PAGE_TIMEOUT = 10000;
```
Comment (rquintana-abstracta) [inline pages/home.page.js]: This is not needed, it's already implemented at BasePage

File: pages/login.page.js [hunk 54]
```diff
@@ -0,0 +1,37 @@
+import BasePage from "./base.page";
+const PAGE_TIMEOUT = 10000;
```
Comment (rquintana-abstracta) [inline pages/login.page.js]: This is not needed, as it's implemented at BasePage

File: pages/logout.page.js [hunk 55]
```diff
@@ -0,0 +1,11 @@
+class LogoutPage extends BasePage {
```
Comment (rquintana-abstracta) [inline pages/logout.page.js:3]: You have to import BasePage in order to use it here

File: tests/login.test.js [hunk 56]
```diff
@@ -0,0 +1,52 @@
+import LoginPage from '../pages/login.page';
+const LOGIN_PAGE = LoginPage;
```
Comment (rquintana-abstracta) [inline tests/login.test.js]: This is not needed, as LoginPage is already an instantiated object here, you can directly use it.

File: tests/logout.test.js [hunk 57]
```diff
@@ -0,0 +1,35 @@
+import LoginPage from '../pages/login.page';
+import LogoutPage from '../pages/logout.page';
+import HomePage from '../pages/home.page';
+import BasePage from '../pages/base.page';
+
+const LOGIN_PAGE = LoginPage;
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: Same comment as above, this is not needed as LoginPage is already an instantiated object, you can directly use it after importing it

File: tests/logout.test.js [hunk 58]
```diff
@@ -0,0 +1,35 @@
+import LoginPage from '../pages/login.page';
+import LogoutPage from '../pages/logout.page';
+import HomePage from '../pages/home.page';
+import BasePage from '../pages/base.page';
```
Comment (rquintana-abstracta) [inline tests/logout.test.js]: You don't need BasePage at this level, only for other pages to inherit from it at their specific pages files

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement new changes

File: pages/home.page.js [hunk 59]
```diff
@@ -0,0 +1,6 @@
+class HomePage extends BasePage{
```
Comment (rquintana-abstracta) [inline pages/home.page.js:2]: Remember to import BasePage un thus file in order to be able to extend from it

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement changes

**Review by rquintana-abstracta — APPROVED** [review]
